agent/registration/csr.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package registration
+
+import (
+	"context"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+
+	certv1 "k8s.io/api/certificates/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	KeySize           = 2048
+	ExpirationSeconds = 86400
+	ByohCSROrg        = "byoh:hosts"
+)
+
+type ByohCSR struct {
+	K8sClient client.Client
+}
+
+func (bcsr *ByohCSR) CreateCSR(hostname, namespace string) (*rsa.PrivateKey, error) {
+	ctx := context.TODO()
+	privKey := &rsa.PrivateKey{}
+	byoCSR := &certv1.CertificateSigningRequest{}
+	err := bcsr.K8sClient.Get(ctx, types.NamespacedName{Name: hostname, Namespace: namespace}, byoCSR)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			klog.Errorf("error getting csr %s in namespace %s, err=%v", hostname, namespace, err)
+			return nil, err
+		}
+		privKey, byoCSR, err = bcsr.generateCSR(hostname, namespace)
+		if err != nil {
+			klog.Errorf("error generating csr %s in namespace %s, err=%v", hostname, namespace, err)
+			return nil, err
+		}
+		err = bcsr.K8sClient.Create(ctx, byoCSR)
+		if err != nil {
+			klog.Errorf("error creating host csr %s in namespace %s, err=%v", hostname, namespace, err)
+			return nil, err
+		}
+	}
+	return privKey, nil
+}
+
+func (bcsr *ByohCSR) generateCSR(hostname, namespace string) (*rsa.PrivateKey, *certv1.CertificateSigningRequest, error) {
+	// Generate Private Key
+	privateKey, err := rsa.GenerateKey(rand.Reader, KeySize)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Generate a new *x509.CertificateRequest template
+	// TODO: validate template
+	csrTemplate := x509.CertificateRequest{
```

Comment:
`csrutils` exists only for testing needs. I thought of keeping the code and test utils separately. WDYS?

Reply:
`builders.go` is again are used for tests. How about `common/utils.go`?

Reply:
> I thought of keeping the code and test utils separately

Agreed. I was suggesting to remove csrutils and put it in builders only.

Reply:
builders can take input for what should be privateKey etc. - if mocked values (like abc, xyz) don't make sense, then these are anyway standard libraries (rsa and x509), so wouldn't be that much of a duplication so that we need to have a common utils IMO.